File: test/test2.c
#define _GNU_SOURCE

#include <sched.h>
#include <sys/wait.h>
#include <stdio.h>
#include <unistd.h>
#include <pthread.h>
#include <string.h>
#include <errno.h>

void *routine(void *arg) {
  printf("thread start %d\n", getpid());
  sleep(10);
  printf("thread end %d\n", getpid());
  return NULL;
}

int main() {
  int err = unshare(CLONE_NEWPID);
  if (err) {
    printf("err = %s", strerror(errno));
    return 1;
  }

  pthread_t tid;
  pthread_create(&tid, NULL, routine, NULL);

  int pid = fork();
  if (pid) {
    int status;
    printf("parent see %d\n", pid);
    pid = waitpid(pid, &status, 0);
    printf("got status %d\n", status);
    printf("result %d\n", pid);
  } else {
    printf("child see %d\n", getpid());
    sleep(3);
  }

  return 0;
}

File: test/test2.rb
# rake compile && sudo ruby -I ./lib ./test/test2.rb

require "runshare"

puts Process.pid
(1..2).each {
  fork {
    parent = Process.pid
    puts "- #{parent}"

    res = RUnshare::unshare(:clone_newpid => true, :fork => true)
    puts "-- unshare=#{res}, pid=#{Process.pid}"

    is_child = parent != Process.pid
    if is_child then
      fork {
        puts "--- #{Process.pid}"
        sleep 10
      }
    end
  }
}

puts 'done'

sleep 10

